In [17]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Otros objetivos
# -----------------------------------------------------------------------
import sys
import os

# Ruta relativa desde notebooks/1-Preprocesing/ a src/
src_path = "../../src/"
sys.path.append(src_path)

import support_comparison as sp  
import pickle
%load_ext autoreload
%autoreload 2

# Visualizaciones
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt

# Para realizar la clasificación y la evaluación del modelo
# -----------------------------------------------------------------------
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, roc_auc_score

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [18]:
df=pd.read_pickle("../../results/cluster_0.pkl")
df

Segment_Consumer  Segment_Corporate  Segment_Home Office  Market_APAC  \
185                   1                  0                    0            0   
197                   1                  0                    0            1   
207                   0                  0                    1            1   
208                   0                  1                    0            0   
214                   0                  0                    1            0   
...                 ...                ...                  ...          ...   
49490                 0                  1                    0            0   
49491                 0                  0                    1            0   
49492                 0                  1                    0            1   
49493                 0                  1                    0            1   
49494                 1                  0                    0            0   

       Market_Africa  Market_Canada  Market_EMEA  Market_EU  Market_LATAM  \
185                0              0            0          1             0   
197                0              0            0          0             0   
207                0              0            0          0             0   
208                0              0            0          0             0   
214                0              0            0          1             0   
...              ...            ...          ...        ...           ...   
49490              1              0            0          0             0   
49491              0              0            1          0             0   
49492              0              0            0          0             0   
49493              0              0            0          0             0   
49494              0              0            0          0             1   

       Market_US  ...  Category_Office Supplies  Category_Technology  \
185            0  ...                         0                    1   
197            0  ...                         1                    0   
207            0  ...                         0                    0   
208            1  ...                         1                    0   
214            0  ...                         0                    0   
...          ...  ...                       ...                  ...   
49490          0  ...                         1                    0   
49491          0  ...                         1                    0   
49492          0  ...                         1                    0   
49493          0  ...                         1                    0   
49494          0  ...                         1                    0   

       Ship Mode  Order Priority  Sales_robust  Quantity_robust  \
185            2               3      6.156041         1.666667   
197            1               2      8.244505         1.000000   
207            2               3      6.798192         1.000000   
208            1               2      7.794999         1.000000   
214            2               2     10.704291         1.333333   
...          ...             ...           ...              ...   
49490          0               1     -0.345552        -0.333333   
49491          2               1     -0.234184         1.000000   
49492          2               2     -0.266836        -0.333333   
49493          2               1     -0.093488         0.666667   
49494          1               2     -0.110387         0.000000   

       Discount_robust  Profit_robust  Shipping Cost_robust  clusters_kmeans  
185                2.5     -34.196691             16.551565                0  
197                2.0      -9.582847             16.215009                0  
207                2.0     -23.549976             16.084715                0  
208                1.0      -9.296161             16.069521                0  
214                0.5      -6.791879             15.917

In [19]:
df2=pd.read_pickle("../../results/cluster_sinencod_0.pkl")
df2

Ship Mode      Segment  Market         Category     Sales  \
185       First Class     Consumer      EU       Technology  1440.840   
197      Second Class     Consumer    APAC  Office Supplies  1900.584   
207          Same Day  Home Office    APAC        Furniture  1582.200   
208      Second Class    Corporate      US  Office Supplies  1801.632   
214       First Class  Home Office      EU        Furniture  2442.069   
...               ...          ...     ...              ...       ...   
49490  Standard Class    Corporate  Africa  Office Supplies     9.612   
49491     First Class  Home Office    EMEA  Office Supplies    34.128   
49492        Same Day    Corporate    APAC  Office Supplies    26.940   
49493        Same Day    Corporate    APAC  Office Supplies    65.100   
49494    Second Class     Consumer   LATAM  Office Supplies    61.380   

       Quantity  Discount    Profit  Shipping Cost Order Priority  \
185           8       0.5 -1268.040         367.35       Critical   
197           6       0.4  -348.516         360.04           High   
207           6       0.4  -870.300         357.21       Critical   
208           6       0.2  -337.806         356.88           High   
214           7       0.1  -244.251         353.57           High   
...         ...       ...       ...            ...            ...   
49490         2       0.7   -21.168           0.02         Medium   
49491         6       0.6   -49.572           0.02         Medium   
49492         2       0.0     1.860           0.01           High   
49493         5       0.0     4.500           0.01         Medium   
49494         3       0.0     1.800           0.00           High   

      clusters_kmeans  
185                 0  
197                 0  
207                 0  
208                 0  
214                 0  
...               ...  
49490               0  
49491               0  
49492               0  
49493               0  
49494               0  

[47098 rows x 11 columns]

In [20]:
df=df.drop(columns=["clusters_kmeans","Profit_robust","Quantity_robust","Discount_robust"])

In [21]:
df["Sales_robust"]=df2["Sales"]

In [22]:
x=df.drop("Sales_robust",axis=1)
y=df[["Sales_robust"]]
x_train, x_test,y_train,y_test=train_test_split(x,y,train_size=0.8,random_state=42)

In [23]:
with open('../../models/model_LinearRegression_cluster_0.pkl', 'rb') as f:
        model_LinearRegression = pickle.load(f)
y_test_pred_model_LinearRegression=model_LinearRegression.predict(x_test)
y_train_pred_model_LinearRegression=model_LinearRegression.predict(x_train)
metricas_model_LinearRegression=sp.metricas(y_train, y_train_pred_model_LinearRegression, y_test, y_test_pred_model_LinearRegression)
metricas_model_LinearRegression

r2_score          MAE           MSE         RMSE
Train                         0.715000    72.213100  16302.536900   127.681400
Test                          0.701200    73.421900  17324.476300   131.622500
Diferencia Train-Test         0.013800    -1.208800  -1021.939400    -3.941100
Porcentaje diferencia (%)     0.487200    -0.415000     -1.519500    -0.759900
Rango valores              2951.155000  2951.155000   2951.155000  2951.155000
Ratio Rango (%)               0.023994     2.467424    569.726314     4.393261
Influencia dif rango (%)      0.000500     0.041000     34.628500     0.133500
Media                       171.514200   171.514200    171.514200   171.514200
Ratio Media(%)                0.412852    42.455668   9802.982260    75.592546
Influencia dif media (%)      0.008000     0.704800    595.833700     2.297800

In [24]:
with open('../../models/model_DecisionTreeRegressor_cluster_0.pkl', 'rb') as f:
        model_DecisionTreeRegressor = pickle.load(f)
y_test_pred_model_DecisionTreeRegressor=model_DecisionTreeRegressor.predict(x_test)
y_train_pred_model_DecisionTreeRegressor=model_DecisionTreeRegressor.predict(x_train)
metricas_model_DecisionTreeRegressor=sp.metricas(y_train, y_train_pred_model_DecisionTreeRegressor, y_test, y_test_pred_model_DecisionTreeRegressor)
metricas_model_DecisionTreeRegressor

r2_score          MAE           MSE         RMSE
Train                         0.821600    49.716100  10204.575100   101.017700
Test                          0.790300    52.557900  12155.719200   110.253000
Diferencia Train-Test         0.031300    -2.841800  -1951.144100    -9.235300
Porcentaje diferencia (%)     0.970900    -1.389300     -4.363000    -2.185700
Rango valores              2951.155000  2951.155000   2951.155000  2951.155000
Ratio Rango (%)               0.027310     1.732779    378.839714     3.579458
Influencia dif rango (%)      0.001100     0.096300     66.114600     0.312900
Media                       171.514200   171.514200    171.514200   171.514200
Ratio Media(%)                0.469903    29.815024   6518.496515    61.589857
Influencia dif media (%)      0.018200     1.656900   1137.599200     5.384600

In [25]:
with open('../../models/model_RandomForestRegressor_cluster_0.pkl', 'rb') as f:
        model_RandomForestRegressor = pickle.load(f)
y_test_pred_model_RandomForestRegressor=model_RandomForestRegressor.predict(x_test)
y_train_pred_model_RandomForestRegressor=model_RandomForestRegressor.predict(x_train)
metricas_model_RandomForestRegressor=sp.metricas(y_train, y_train_pred_model_RandomForestRegressor, y_test, y_test_pred_model_RandomForestRegressor)
metricas_model_RandomForestRegressor

r2_score          MAE           MSE         RMSE
Train                         0.855900    44.848200   8245.390000    90.804100
Test                          0.803700    50.475000  11378.983100   106.672300
Diferencia Train-Test         0.052200    -5.626800  -3133.593100   -15.868200
Porcentaje diferencia (%)     1.572700    -2.951400     -7.983900    -4.017700
Rango valores              2951.155000  2951.155000   2951.155000  2951.155000
Ratio Rango (%)               0.028118     1.615015    332.486316     3.345748
Influencia dif rango (%)      0.001800     0.190700    106.181900     0.537700
Media                       171.514200   171.514200    171.514200   171.514200
Ratio Media(%)                0.483808    27.788720   5720.917889    57.568528
Influencia dif media (%)      0.030400     3.280700   1827.016700     9.251800

In [26]:
with open('../../models/model_GradientBoostingRegressor_cluster_0.pkl', 'rb') as f:
        modelo_gradient_boosting = pickle.load(f)
y_test_pred_modelo_gradient_boosting=modelo_gradient_boosting.predict(x_test)
y_train_pred_modelo_gradient_boosting=modelo_gradient_boosting.predict(x_train)
metricas_modelo_gradient_boosting=sp.metricas(y_train, y_train_pred_modelo_gradient_boosting, y_test, y_test_pred_modelo_gradient_boosting)
metricas_modelo_gradient_boosting

r2_score          MAE           MSE         RMSE
Train                         0.842700    46.914100   8998.265400    94.859200
Test                          0.806900    50.204600  11196.157500   105.811900
Diferencia Train-Test         0.035800    -3.290500  -2197.892100   -10.952700
Porcentaje diferencia (%)     1.085100    -1.694100     -5.441800    -2.729000
Rango valores              2951.155000  2951.155000   2951.155000  2951.155000
Ratio Rango (%)               0.027948     1.645435    342.144396     3.399874
Influencia dif rango (%)      0.001200     0.111500     74.475700     0.371100
Media                       171.514200   171.514200    171.514200   171.514200
Ratio Media(%)                0.480893    28.312146   5887.099406    58.499850
Influencia dif media (%)      0.020900     1.918500   1281.463600     6.385900

In [27]:
lista_nombres_modelos=["modelo_linear_regression","modelo_decision_tree","modelo_random_forest","modelo_gradient_boosting",]

In [28]:
sp.combinar_metricas(lista_nombres_modelos,metricas_model_LinearRegression,metricas_model_DecisionTreeRegressor ,metricas_model_RandomForestRegressor ,metricas_modelo_gradient_boosting)

Train        Test  Diferencia Train-Test  \
r2_score      0.7150      0.7012                 0.0138   
MAE          72.2131     73.4219                -1.2088   
MSE       16302.5369  17324.4763             -1021.9394   
RMSE        127.6814    131.6225                -3.9411   
r2_score      0.8216      0.7903                 0.0313   
MAE          49.7161     52.5579                -2.8418   
MSE       10204.5751  12155.7192             -1951.1441   
RMSE        101.0177    110.2530                -9.2353   
r2_score      0.8559      0.8037                 0.0522   
MAE          44.8482     50.4750                -5.6268   
MSE        8245.3900  11378.9831             -3133.5931   
RMSE         90.8041    106.6723               -15.8682   
r2_score      0.8427      0.8069                 0.0358   
MAE          46.9141     50.2046                -3.2905   
MSE        8998.2654  11196.1575             -2197.8921   
RMSE         94.8592    105.8119               -10.9527   

          Porcentaje diferencia (%)  Rango valores  Ratio Rango (%)  \
r2_score                     0.4872       2951.155         0.023994   
MAE                         -0.4150       2951.155         2.467424   
MSE                         -1.5195       2951.155       569.726314   
RMSE                        -0.7599       2951.155         4.393261   
r2_score                     0.9709       2951.155         0.027310   
MAE                         -1.3893       2951.155         1.732779   
MSE                         -4.3630       2951.155       378.839714   
RMSE                        -2.1857       2951.155         3.579458   
r2_score                     1.5727       2951.155         0.028118   
MAE                         -2.9514       2951.155         1.615015   
MSE                         -7.9839       2951.155       332.486316   
RMSE                        -4.0177       2951.155         3.345748   
r2_score                     1.0851       2951.155         0.027948   
MAE                         -1.6941       2951.155         1.645435   
MSE                         -5.4418       2951.155       342.144396   
RMSE                        -2.7290       2951.155         3.399874   

          Influencia dif rango (%)     Media  Ratio Media(%)  \
r2_score                    0.0005  171.5142        0.412852   
MAE                         0.0410  171.5142       42.455668   
MSE                        34.6285  171.5142     9802.982260   
RMSE                        0.1335  171.5142       75.592546   
r2_score                    0.0011  171.5142        0.469903   
MAE                         0.0963  171.5142       29.815024   
MSE                        66.1146  171.5142     6518.496515   
RMSE                        0.3129  171.5142       61.589857   
r2_score                    0.0018  171.5142        0.483808   
MAE                         0.1907  171.5142       27.788720   
MSE                       106.1819  171.5142     5720.917889   
RMSE                        0.5377  171.5142       57.568528   
r2_score                    0.0012  171.5142        0.480893   
MAE                         0.1115  171.5142       28.312146   
MSE                        74.4757  171.5142     5887.099406   
RMSE                        0.3711  171.5142       58.499850   

          Influencia dif media (%)                    Modelo  
r2_score                    0.0080  modelo_linear_regression  
MAE                         0.7048  modelo_linear_regression  
MSE                       595.8337  modelo_linear_regression  
RMSE                        2.2978  modelo_linear_regression  
r2_score                    0.0182      modelo_decision_tree  
MAE                         1.6569      modelo_decision_tree  
MSE                      1137.5992      modelo_decision_tree  
RMSE                        5.3846      modelo_decision_tree  
r2_score                    0.0304      modelo_random_forest  
MAE                         3.2807      modelo_random_forest  
MSE                  

1. Rendimiento General (R² Score):

Mejor modelo: Random Forest con un R² en entrenamiento (85.59%) y prueba (80.37%). Aunque el Gradient Boosting también tiene un R² alto (84.27% en entrenamiento y 80.69% en prueba), el Random Forest logra un rendimiento ligeramente superior en prueba.
Modelo más consistente: Linear Regression tiene la menor diferencia entre entrenamiento y prueba (0.4872%), lo que indica una buena generalización, aunque su R² es el más bajo.

2. Error Absoluto Medio (MAE):

Mejor modelo: Random Forest con el MAE más bajo en ambos conjuntos (44.85 en entrenamiento y 50.48 en prueba), indicando errores absolutos menores.
Peor modelo: Linear Regression, con MAE significativamente mayor (72.21 en entrenamiento y 73.42 en prueba).

3. Error Cuadrático Medio (MSE) y Raíz Cuadrada del Error Cuadrático Medio (RMSE):

Mejor modelo en MSE y RMSE: Random Forest, con los valores más bajos de MSE (8245.39 en entrenamiento y 11378.98 en prueba) y RMSE (90.80 en entrenamiento y 106.67 en prueba). Esto refleja una menor amplificación de errores.
Peor modelo en MSE y RMSE: Linear Regression, con los errores más altos en ambos conjuntos.

4. Generalización y Diferencias Train-Test:

Más consistente: Gradient Boosting, con diferencias moderadas en MAE, MSE y RMSE entre los conjuntos (MAE: -1.69%, MSE: -5.44%, RMSE: -2.73%), indicando una buena capacidad de generalización.
Modelo con más diferencias: Random Forest, con mayores diferencias en MSE y RMSE (MSE: -7.98%, RMSE: -4.02%), lo que podría indicar un ligero sobreajuste.

5. Relación con el Rango de Valores:

Todos los modelos tienen un bajo impacto del R² respecto al rango de valores (<0.03%), pero el Linear Regression amplifica significativamente los errores en relación al rango, como lo muestra su alto ratio para el MSE (569.73%).

Recomendación

- Modelo recomendado: Random Forest debido a su excelente rendimiento en R², MAE, MSE y RMSE. Aunque tiene un ligero sobreajuste, es el modelo más robusto para predicciones precisas.
- Alternativa: Gradient Boosting combina un rendimiento sólido y una mejor generalización, siendo ideal si la consistencia es prioritaria.